File: guild.go
package discord

import "fmt"

// guild struct
type Guild struct {
    Id                          string              `json:"id"`
    Name                        string              `json:"name"`
    SafetyAlertsChannel         *Channel            `json:"_"`
    GuildScheduledEvents        []any               `json:"guild_scheduled_events"` // add later
    Members                     []any               `json:"_"` // []*Member
    Splash                      string              `json:"splash"`
    MaxVideoChannelUsers        int                 `json:"max_video_channel_users"`
    MfaLevel                    int                 `json:"mfa_level"`
    VanityUrlCode               string              `json:"vanity_url_code"`
    Stickers                    []any               `json:"stickers"` // []*Sticker
    ActivityInstances           map[string]any      `json:"activity_instances"`
    EmbeddedActivities          []any               `json:"embedded_activities` // add later
    ApplicationCommandCounts    map[string]any      `json:"application_command_counts"` // add later
    MaxStageVideoChannelUsers   int                 `json:"max_stage_video_channel_users"`
    DiscoverySplash             string              `json:"discovery_splash"`
    LatestOnboardingQuestionId  string              `json:"latest_onboarding_question_id`
    PublicUpdatesChannel        *Channel            `json:"_"`
    SystemChannelFlags          int                 `json:"system_channel_flags"`
    InventorySettings           any                 `json:"inventory_settings"` // add later
    Presences                   []any               `json:"presences"` // add later 
    Features                    []string            `json:"features"`
    PremiumProgressBarEnabled   bool                `json:"premium_progress_bar_enabled"`
    SystemChannel               *Channel            `json:"_"`
    Threads                     any                 `json:"_"`
    Channels                    channelManager      `json:"_"`
    PremiumSubscriptionCount    int                 `json:"premium_subscription_count"`
    DefaultMessageNotifications int                 `json:"default_message_notifications"`
    Description                 string              `json:"descritpion"`
    Unavailable                 bool                `json:"unavailable"`
    Banner                      string              `json:"banner"`
    VerificationLevel           int                 `json:"verification_level"`
    Emojis                      []any               `json:"_"` // []*Emoji
    MemberCount                 int                 `json:"member_count"`
    AfkTimeout                  int                 `json:"aft_timeout"`
    ExplicitContentFilter       int                 `json:"explicit_conten_filter"`
    AfkChannel                  *Channel            `json:"_"`
    SoundboardSounds            []any               `json:"soundboard_sounds"` // add later
    NsfwLevel                   int                 `json:"nsfw_level"`
    PreferredLocale             string              `json:"preferred_locale"`
    ApplicationId               string              `json:"application_id"`
    Nsfw                        bool                `json:"nsfw"`
    VoiceStates                 []any               `json:"voice_states"` // add later
    Icon                        string              `json:"icon"`
    Large                       bool                `json:"large"`
    PremiumTier                 int                 `json:"premium_tier"`
    MaxMembers                  int                 `json:"max_members"`
    Roles                       []*Role             `json:"_"`
    JoinedAt                    string              `json:"joined_at"`
    HomeHeader                  string              `json:"home_header"`
    Owner                       string              `json:"_"` // *Member
    Version                     int                 `json:"version"`
    IncidentsData               any                 `json:"incidents_data"` // add later
    Clan                        any                 `json:"clan"` // add later
    Lazy                        bool                `json:"lazy"`
    RulesChannel                *Channel            `json:"_"`
    StageInstances              []any               `json:"stage_instances"` // add later
    HubType                     any                 `json"hub_type"` // add later
}
// guild manager
type guildManager struct { guilds map[string]*Guild }
func (manager *guildManager) PrintAll() {
    for _, g := range manager.guilds {
        fmt.Printf("%#v\n", g)
    }
}
func (manager *guildManager) Size() int {
    return len(manager.guilds)
}
func (manager *guildManager) Add(guild *Guild) {
    manager.guilds[guild.Id] = guild
}
func (manager *guildManager) Get(guild_id string) *Guild {
    return manager.guilds[guild_id]
}
func (manager *guildManager) Replace(guild_id string, new_guild *Guild) {
    manager.guilds[guild_id] = new_guild
}
func (manager *guildManager) Remove(guild_id string) {
    delete(manager.guilds, guild_id)
}
func (manager *guildManager) Fetch(guild_id string) {
  
}

File: src/guild.rs
use crate::manager;
use crate::channel;
use crate::member;

#[derive(Debug)]
pub struct Guild<'a> {
    pub id:                                     String,
    // name

    // joined_at
    // large
    // unavailable
    // member_count
    // voice_states
    pub members:                                manager::Manager<'a, member::Member<'a>>,
    pub channels:                               manager::Manager<'a, channel::Channel<'a>>, 
    // threads
    // presences
    // stage_instances
    // guild_scheduled_events

    // icon
    // icon_hash
    // splash
    // discovery_splash
    // owner
    // afk_channel *
    // afk_timeout
    // widget_enabled
    // widget_channel *
    pub verification_level:                     GuildVerificationLevel,
    pub default_message_notifications_level:    GuildDefaultMessageNotificationsLevel,
    pub explicit_content_filter_level:          GuildExplicitContentFilterLevel,
    // roles (manager)
    // emojis (manager)
    pub features:                               Vec<GuildFeature>,
    pub mfa_level:                              GuildMfaLevel,
    // application
    // system_channel *
    // system_channel_flags (might be optional)
    // rules_channel *
    // max_presences
    // vanity_url
    // description
    // banner
    pub premium_tier:                           GuildPremiumTier,
    // premium_subscription_count
    pub preferred_locale:                       GuildLocale,
    // public_updates_channel *
    // max_video_channel_users
    // max_stage_video_channel_users
    // approximate_member_count
    // approximate_presence_count
    // welcome_screen
    pub nsfw_level:                             GuildNsfwLevel
    // stickers (manager)
    // premium_progress_bar_enabled
    // safety_alerts_channel *
}


#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum GuildVerificationLevel {
    NONE,
    LOW,
    MEDIUM,
    HIGH,
    VERY_HIGH
}

#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum GuildDefaultMessageNotificationsLevel {
    ALL_MESSAGES,
    ONLY_MENTIONS
}

#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum GuildExplicitContentFilterLevel {
    DISABLED,
    MEMBERS_WITHOUT_ROLES,
    ALL_MEMBERS
}

#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum GuildFeature {
    ANIMATED_BANNER,
    ANIMATED_ICON,
    APPLICATION_COMMAND_PERMISSIONS_V2,
    AUTO_MODERATION,
    BANNER,
    COMMUNITY,
    CREATOR_MONETIZABLE_PROVISIONAL,
    CREATOR_STORE_PAGE,
    DEVELOPER_SUPPORT_SERVER,
    DISCOVERABLE,
    FEATURABLE,
    INVITES_DISABLED,
    INVITE_SPLASH,
    MEMBER_VERIFICATION_GATE_ENABLED,
    MORE_STICKERS,
    NEWS,
    PARTNERED,
    RAID_ALERTS_DISABLED,
    ROLE_ICONS,
    ROLE_SUBSCRIPTIONS_AVAILABLE_FOR_PURCHASE,
    ROLE_SUBSCRIPTIONS_ENABLED,
    TICKETED_EVENTS_ENABLED,
    VANITY_URL,
    VERIFIED,
    VIP_REGIONS,
    WELCOME_SCREEN_ENABLED
}

#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum GuildMfaLevel {
    NONE,
    ELEVATED
}

#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum GuildPremiumTier {
    NONE,
    TIER_1,
    TIER_2,
    TIER_3
}

#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum GuildLocale {
    INDONESIAN,
    DANISH,
    GERMAN,
    ENGLISH_UK,
    ENGLISH_US,
    SPANISH,
    SPANISH_LATAM,
    FRENCH,
    CROATIAN,
    ITALIAN,
    LITHUANIAN,
    HUNGARIAN,
    DUTCH,
    NORWEGIAN,
    POLISH,
    PORTUGUESE,
    ROMANIAN,
    FINNISH,
    SWEDISH,
    VIETNAMESE,
    TURKISH,
    CZECH,
    GREEK,
    BULGARIAN,
    RUSSIAN,
    UKRAINIAN,
    HINDI,
    THAI,
    CHINESE_CN,
    JAPANESE,
    CHINESE_TW,
    KOREAN
}

#[derive(Debug)]
#[allow(non_camel_case_types)]
pub enum GuildNsfwLevel {
    DEFAULT,
    EXPLICIT,
    SAFE,
    AGE_RESTRICTED
}